### ISSUE TITLE
404 Not Found during Build w/ dependent repositories

### ISSUE THREAD

**Issue by buzzwick** [microsoft/AL-Go#791]
Getting a 404 Not Found when looking for the dependent repository against which the subject repository is built. Simply cannot figure out where the error is such that the repo is not found:

Run try { D:\a\_actions\microsoft\AL-Go-Actions\v3.0\RunPipeline/RunPipeline.ps1 -actor $ENV:_actor -token $ENV:_token -parentTelemetryScopeJson $ENV:_parentTelemetryScopeJson -project $ENV:_project -projectDependenciesJson $ENV:_projectDependenciesJson -settingsJson $ENV:_settingsJson -secretsJson $ENV:_secretsJson -buildMode $ENV:_buildMode } catch { Write-Host "::Error::Unexpected error when running action ($($_.Exception.Message.Replace("`r",'').Replace("`n",' ')))"; exit 1 }
Downloading BcContainerHelper latest version from CDN
BcContainerHelper version 6.0.0
BC.HelperFunctions emits usage statistics telemetry to Microsoft
use settings and secrets
Analyzing repository
Checking project dependencies
Checking appDependencyProbingPaths
Dependency to projects '*' in https://github.com/CAI-Software-Inc / Precious-Metals-Solution@main, version latest, release status release
Warning: No performance test apps found in bcptTestFolders in .AL-Go\settings.json
Warning: No test apps found in testFolders in .AL-Go\settings.json
Project: 
Downloading dependencies
  Locating all Apps artifacts from probing paths
  Analyzing releases https://api.github.com/repos/CAI-Software-Inc%20/%20Precious-Metals-Solution/releases
  Error: The remote server returned an error: (404) Not Found. 
  
  Not Found
  
  Error: RunPipeline action failed. Error: The remote server returned an error: (404) Not Found.   Not Found  Stacktrace: at InvokeWebRequest, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\Github-Helper.psm1: line 91 at GetReleases, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\Github-Helper.psm1: line 453 at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\Github-Helper.psm1: line 171 at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\Github-Helper.psm1: line 111 at Get-dependencies, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\Github-Helper.psm1: line 108 at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v3.0\RunPipeline\RunPipeline.ps1: line 136 at <ScriptBlock>, D:\a\_temp\d3a1e8a8-f0da-4dcf-b733-0596dabac635.ps1: line 2 at <ScriptBlock>, <No file>: line 1
  AL-Go action ran: RunPipeline Telemetry Correlation Id: 63409951-1fdc-4c5e-9b0d-5dce62e08132
  Removing BcContainerHelper
  Error: Process completed with exit code 1.

**Comment by buzzwick** [comment]
Now we are getting the same Not Found message during a build of the main repository as well:
No performance test apps found in bcptTestFolders in .AL-Go\settings.json
Checking appDependencyProbingPaths
Dependency to projects '*' in https://github.com/CAI-Software-Inc / Precious-Metals-Solution@main, version latest, release status release
Using secret GhTokenWorkflow for access to repository
Locating Apps artifacts for projects: *
Analyzing releases https://api.github.com/repos/CAI-Software-Inc%20/%20Precious-Metals-Solution/releases
Error: The remote server returned an error: (404) Not Found.

**Comment by buzzwick** [comment]
Running the Action Update AL-Go System Files, we get more information:

Unexpected error when running action. Error Message: Failed to update AL-Go System Files. Make sure that the personal access token, defined in the secret called GhTokenWorkflow, is not expired and it has permission to update workflows. (Error was Cloning into 'Precious-Metals-Solution'... fatal: unable to access '***github.com/CAI-Software-Inc/Precious-Metals-Solution/': URL rejected: **Port number was not a decimal number between 0 and 65535**  ExitCode: 128 Commandline: git clone "***github.com/CAI-Software-Inc/Precious-Metals-Solution" ), StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v3.3\CheckForUpdates\CheckForUpdates.ps1: line 523 <- at <ScriptBlock>, D:\a\_temp\868b2193-eaa9-4382-8e4e-c8f6f018b12f.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1

This is _after_ I had updated the GHTOKENWORKFLOW secret with a valid value. I have no idea which port number the message is referring to, or where it would be found.

**Comment by buzzwick** [comment]
The GHTOKENWORKFLOW value was not valid; I had put in an authcontext for the environment by mistake. Fine-grained PATs apparently do not work either. But a classic PAT does. We still get a Not Found error however, when looking for the dependent repository:

Data-Migration/.dependencies
Downloading project dependencies from current build
Downloading project dependencies from probing paths
Analyzing repository
No performance test apps found in bcptTestFolders in .AL-Go\settings.json
Checking appDependencyProbingPaths
Dependency to projects '*' in https://github.com/CAI-Software-Inc / Precious-Metals-Solution@main, version latest, release status release
Using secret GhTokenWorkflow for access to repository
Locating Apps artifacts for projects: *
Analyzing releases https://api.github.com/repos/CAI-Software-Inc%20/%20Precious-Metals-Solution/releases
Error: The remote server returned an error: (404) Not Found. 

Not Found

Error: Unexpected error when running action. Error Message: The remote server returned an error: (404) Not Found.   Not Found , StackTrace: at InvokeWebRequest, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\Github-Helper.psm1: line 117 <- at GetReleases, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\Github-Helper.psm1: line 495 <- at GetDependencies, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\Github-Helper.psm1: line 201 <- at DownloadDependenciesFromProbingPaths, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\DownloadProjectDependencies\DownloadProjectDependencies.Action.ps1: line 16 <- at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\DownloadProjectDependencies\DownloadProjectDependencies.Action.ps1: line 96 <- at <ScriptBlock>, D:\a\_temp\311d683a-28b6-4[87](https://github.com/CAI-Software-Inc/Johnson-Matthey-Data-Migration/actions/runs/6695332661/job/18190709769#step:7:92)6-a4a7-20d4ce66ea29.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1
Error: Process completed with exit code 1.

**Comment by freddydk** [comment]
Looking at the output, I see this:

![image](https://github.com/microsoft/AL-Go/assets/10775043/2a77fb2b-5f1d-4ce7-be13-f678356f7081)

the "https://" is replaced by three *'s

That normally happens if a secret has that exact value.

When there is a secret with a value like this, then that value becomes invalid and thus is removed from the values transferred from one job to another.

Could it be that one of your authcontext secrets has a property with this value or maybe a url with https:// and then a space or like???

If you remove the repo secrets, I think it will work - and then you can add them back in to figure out which one is bad.
Sorry, but I do not have a good way to identify "bad" secrets.

**Comment by buzzwick** [comment]
Okay. At least it is something to try...

**Comment by buzzwick** [comment]
This seems to be a configuration error, but where exactly?
I believe the 404 Not Found is resulting from the named dependent repo not being found using the URL coming from ALGOORGSETTINGS:  "appDependencyProbingPaths":[{"repo":"CAI-Software-Inc/Precious-Metals-Solution","release_status":"latest","AuthTokenSecret": "GhTokenWorkflow"}]

Testing in Postman using the same api URIs used by the workflow, we find that

`https://api.github.com/orgs/CAI-Software-Inc`

generates the organization metadata:

{
    "login": "CAI-Software-Inc",
    "id": 111072522,
    "node_id": "O_kgDOBp7VCg",
    "url": "https://api.github.com/orgs/CAI-Software-Inc",
    "repos_url": "https://api.github.com/orgs/CAI-Software-Inc/repos",
    "events_url": "https://api.github.com/orgs/CAI-Software-Inc/events",
    "hooks_url": "https://api.github.com/orgs/CAI-Software-Inc/hooks",
    "issues_url": "https://api.github.com/orgs/CAI-Software-Inc/issues",
    "members_url": "https://api.github.com/orgs/CAI-Software-Inc/members{/member}",
    "public_members_url": "https://api.github.com/orgs/CAI-Software-Inc/public_members{/member}",
    "avatar_url": "https://avatars.githubusercontent.com/u/111072522?v=4",
    "description": "",
    "name": "CAI Software Inc.",
    "company": null,
    "blog": null,
    "location": null,
    "email": null,
    "twitter_username": null,
    "is_verified": false,
    "has_organization_projects": true,
    "has_repository_projects": true,
    "public_repos": 0,
    "public_gists": 0,
    "followers": 1,
    "following": 0,
    "html_url": "https://github.com/CAI-Software-Inc",
    "created_at": "2022-08-11T15:10:43Z",
    "updated_at": "2023-10-20T19:48:29Z",
    "archived_at": null,
    "type": "Organization"
}

Adding the /repos segment at the end

`https://api.github.com/orgs/CAI-Software-Inc/repos`

generates only the empty set:

`[]`

with no error.  I am using a PAT with access to all repositories for this organization.

Removing the appDependencyProbingPaths from ALGOORGSETTINGS eliminated the Not Found error, but instead we get

`Run microsoft/AL-Go-Actions/Deploy@v4.0
Run $errorActionPreference = "Stop"; $ProgressPreference = "SilentlyContinue"; Set-StrictMode -Version 2.0
Downloading BcContainerHelper latest version from CDN
Import from C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\BcContainerHelper.ps1
BcContainerHelper version 6.0.0
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Analyzing artifacts matching *-main-*-*
https://api.github.com/repos/CAI-Software-Inc/Precious-Metals-Solution/actions/artifacts?per_page=100&page=1
Matching artifacts:
- No matching artifacts found
**Analyzing artifacts matching *-main-*-*
https://api.github.com/repos/CAI-Software-Inc/Precious-Metals-Solution/actions/artifacts?per_page=100&page=1
Matching artifacts:
- No matching artifacts found**
AL-Go action ran: Deploy Telemetry Correlation Id: 3210aca1-e235-4139-b0d5-75e7[44](https://github.com/CAI-Software-Inc/Precious-Metals-Solution/actions/runs/6712171307/job/18241164468#step:6:46)c7cb81
**Unexpected error when running action. Error Message: Could not find any Apps artifacts for projects *, version latest**, StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\Deploy\Deploy.ps1: line 142 <- at <ScriptBlock>, D:\a\_temp\6f62e9de-a77e-4326-957f-10ba4d959dfe.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1`

So there are two questions: why no access to the repositories, and why not finding any artifacts for the latest release (because they definitely exist, and match the

**Comment by AskeHolst** [comment]
We have the same issue in all repos using appDependencyProbingPaths, it started appearing yesterday I believe.

The workaround now is switching to downloading latest from the artifacts delivered to AZ storage using "installApps", which works fine.

**Comment by freddydk** [comment]
I issued a hotfix yesterday for #799 - I think this is the same problem, a regression in the GetArtifacts function.
2nd part of the hotfix should be in within the next hours.

**Comment by freddydk** [comment]
This should now be fixed, sorry for the inconvenience

**Comment by buzzwick** [comment]
We may need some additional illumination around the GHTOKENWORKFLOW variable. We were able to update the AL-Go system files in our "main" repository without error, but in another repo with a dependency to the main one, we get

`Error: Unexpected error when running action. Error Message: Failed to update AL-Go System Files. Make sure that the personal access token, defined in the secret called GhTokenWorkflow, is not expired and it has permission to update workflows. (Error was Cloning into 'Johnson-Matthey-UK'... remote: Write access to repository not granted. fatal: unable to access 'https://github.com/CAI-Software-Inc/Johnson-Matthey-UK/': The requested URL returned error: 403  ExitCode: 128 Commandline: git clone ***github.com/CAI-Software-Inc/Johnson-Matthey-UK ), StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\CheckForUpdates\CheckForUpdates.ps1: line 523 <- at <ScriptBlock>, D:\a\_temp\f7759cee-65e1-4043-aaa1-5498286637f9.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1`


using the same PAT that is in place in GHTOKENWORKFLOW, we do not seem to have access to the GitHub repositories through the api: 

`https://api.github.com/orgs/CAI-Software-Inc/repos`
generates only the empty set:
[]

No error, but no repositories. 

Shouldn't both updates fail?

**Comment by freddydk** [comment]
The above dump comes from using directCommit - do you maybe have branch rules in place saying you cannot use direct commit?

**Comment by buzzwick** [comment]
No branch rules in place. Tried without directCommit:

Unexpected error when running action. Error Message: Failed to create a pull-request to AL-Go System Files. Make sure that the personal access token, defined in the secret called GhTokenWorkflow, is not expired and it has permission to update workflows. (Error was Cloning into 'Johnson-Matthey-UK'... remote: Write access to repository not granted. fatal: unable to access 'https://github.com/CAI-Software-Inc/Johnson-Matthey-UK/': The requested URL returned error: 403  ExitCode: 128 Commandline: git clone ***github.com/CAI-Software-Inc/Johnson-Matthey-UK ), StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\CheckForUpdates\CheckForUpdates.ps1: line 526 <- at <ScriptBlock>, D:\a\_temp\141cac32-2ba2-48ac-82d8-c6430fd7b576.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1

I can try re-generating the PAT (again), but it seems to work on other repos in the same org. And it was created with access to all repositories.

**Comment by freddydk** [comment]
The PAT is not a fine grained PAT right? (they don't work)

**Comment by buzzwick** [comment]
It is not fine-grained any more. I am pretty sure I tried this both ways, but I created a new Classic PAT and now we seem able to move on to the next obstacle: "Illegal characters in path."
Here is the log from the last run of Test Current Version, which I ran to force a build (is there a better way to do that?):

Run microsoft/AL-Go-Actions/RunPipeline@v4.0
Run $errorActionPreference = "Stop"; $ProgressPreference = "SilentlyContinue"; Set-StrictMode -Version 2.0
Import from C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\BcContainerHelper.ps1
BcContainerHelper version 6.0.0
BC.HelperFunctions emits usage statistics telemetry to Microsoft
use settings and secrets
Analyzing repository
No performance test apps found in bcptTestFolders in .AL-Go\settings.json
Warning: No test apps found in testFolders in .AL-Go\settings.json
Checking appDependencyProbingPaths
Warning: Using the legacy CodeSignCertificateUrl and CodeSignCertificatePassword parameters. Consider using the new Azure Keyvault signing instead. Go to https://aka.ms/ALGoSettings#keyVaultCodesignCertificateName to find out more
Locating previous release
Invoke Run-AlPipeline with buildmode Default
Parameters
Creating container
Resolving dependencies
Compiling apps
Removing container
Run-AlPipeline Telemetry Correlation Id: 1744fab5-dbbb-4486-9b9e-32b547d99a73
AL-Go action ran: RunPipeline Telemetry Correlation Id: 6532557b-e39f-4eb0-9aea-9fce9f4c1085
Get Event Log from container
Getting event log for bc6773960762
Error getting event log from container: Cannot overwrite the item D:\a\Johnson-Matthey-UK\Johnson-Matthey-UK\ContainerEventLog.evtx with itself.
**Error: Unexpected error when running action. Error Message: Exception calling "DownloadFile" with "2" argument(s): "Illegal characters in path.",** StackTrace: at Compile-AppInBcContainer, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Compile-AppInNavContainer.ps1: line 486 <- at <ScriptBlock>, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 779 <- at <ScriptBlock>, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 1894 <- at <ScriptBlock>, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 1441 <- at <ScriptBlock>, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 1434 <- at <ScriptBlock>, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 904 <- at Run-AlPipeline, C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\AppHandling\Run-AlPipeline.ps1: line 866 <- at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\RunPipeline\RunPipeline.ps1: line 371 <- at <ScriptBlock>, D:\a\_temp\3c33174d-bbe9-4615-8cfa-b1[58](https://github.com/CAI-Software-Inc/Johnson-Matthey-UK/actions/runs/6773960762/job/18410015697#step:8:60)e65b39d0.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1
Error: Process completed with exit code 1.

**Comment by buzzwick** [comment]
Now back to "no matching artifacts found":

Run microsoft/AL-Go-Actions/Deploy@v4.0
Run $errorActionPreference = "Stop"; $ProgressPreference = "SilentlyContinue"; Set-StrictMode -Version 2.0
Downloading BcContainerHelper latest version from CDN
Import from C:\ProgramData\BcContainerHelper\6.0.0\BcContainerHelper\BcContainerHelper.ps1
BcContainerHelper version 6.0.0
BC.HelperFunctions emits usage statistics telemetry to Microsoft
Analyzing artifacts matching *-main-*-l
https://api.github.com/repos/CAI-Software-Inc/Precious-Metals-Solution/actions/artifacts?per_page=100&page=1
No matching buildOutput artifacts found
Analyzing artifacts matching *-main-*-l
https://api.github.com/repos/CAI-Software-Inc/Precious-Metals-Solution/actions/artifacts?per_page=100&page=1
No matching buildOutput artifacts found
AL-Go action ran: Deploy Telemetry Correlation Id: 4526032b-fa9f-43e9-8f10-96672bfbc165
Error: Unexpected error when running action. Error Message: Could not find any Apps artifacts for projects *, version l, StackTrace: at <ScriptBlock>, D:\a\_actions\microsoft\AL-Go-Actions\v4.0\Deploy\Deploy.ps1: line 1[42](https://github.com/CAI-Software-Inc/Precious-Metals-Solution/actions/runs/6812619784/job/18525391306#step:6:44) <- at <ScriptBlock>, D:\a\_temp\2967c8d4-f5c0-[43](https://github.com/CAI-Software-Inc/Precious-Metals-Solution/actions/runs/6812619784/job/18525391306#step:6:45)19-bd85-6eaca16157de.ps1: line 4 <- at <ScriptBlock>, <No file>: line 1
Error: Process completed with exit code 1.

The only artifacts I am aware of are those associated with our latest release: 
Precious-Metals-Solution-main-Apps-22.2.47.0.zip

**Comment by freddydk** [comment]
It looks like you specified a lowercase L in the version field - and it is trying to locate artifacts with this version (which cannot be found).
I can see you ran the Publish again shortly after and this time it worked.

**Comment by buzzwick** [comment]
I was able to publish on a successive attempt. I think I know where the "l" came from: relying on autocomplete in the Version field, but getting only the leading character. Should have been "latest." Never would have caught that, thanks.

**Comment by buzzwick** [comment]
Absolutely clear! I just meant I never would have made the connection to where that "version" came from. Soon as you mentioned it I knew what had happened.,

**Comment by buzzwick** [comment]
Going to need some further troubleshooting help. When I run Publish to Environment, I get
`Could not find any Apps artifacts for projects *, version latest`

When I run CI/CD to generate some Apps artifacts, I get
`The remote server returned an error: (404) Not Found. `

This feels like a Catch-22, and it would help to know what is Not Found, and how to get some artifacts in place.

**Comment by freddydk** [comment]
Which repo is this?

**Comment by buzzwick** [comment]
CAI-Software-Inc/Elemet-Inc

**Comment by buzzwick** [comment]
Thank you! Yes those are normal compiler errors--for us. I think I removed AuthTokenSecret when I removed the whole appDependencyProbingPaths, then only put back the path part. 
I will try this on our other repo with the same problem.

Should the Workshop slide include the AuthTokenSecret pair as well? Right now it only has this:

"Now your organization variable ALGoOrgSettings, and add:

`"appDependencyProbingPaths": [
        {
            "repo": "freddydkorg/Common",
            "release_status": "latestBuild"
        }
    ]`
